File: kora-app-annotation-processor/src/main/java/ru/tinkoff/kora/kora/app/annotation/processor/declaration/ComponentDeclaration.java
package ru.tinkoff.kora.kora.app.annotation.processor.declaration;

import com.palantir.javapoet.ClassName;
import com.palantir.javapoet.CodeBlock;
import jakarta.annotation.Nullable;
import ru.tinkoff.kora.annotation.processor.common.*;
import ru.tinkoff.kora.kora.app.annotation.processor.ProcessingContext;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.ExtensionResult;

import javax.lang.model.element.*;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeMirror;
import javax.tools.Diagnostic;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.function.Function;

public sealed interface ComponentDeclaration {
    TypeMirror type();

    Element source();

    @Nullable
    String tag();

    default boolean isTemplate() {
        return TypeParameterUtils.hasTypeParameter(this.type());
    }

    default boolean isDefault() {
        return false;
    }

    boolean isInterceptor();

    String declarationString();

    record FromModuleComponent(TypeMirror type, ModuleDeclaration module, @Nullable String tag, ExecutableElement method, List<TypeMirror> methodParameterTypes,
                               List<TypeMirror> typeVariables, boolean isInterceptor) implements ComponentDeclaration {
        @Override
        public Element source() {
            return this.method;
        }

        @Override
        public boolean isDefault() {
            return AnnotationUtils.findAnnotation(this.method, CommonClassNames.defaultComponent) != null;
        }

        @Override
        public String declarationString() {
            return module.element().getQualifiedName() + "." + method.getSimpleName();
        }

        @Override
        public String toString() {
            final StringBuilder sb = new StringBuilder("FromModuleComponent[");
            sb.append("type=").append(type);
            sb.append(", module=").append(module);
            sb.append(", method=").append(method);
            if (tag != null) {
                sb.append(", tag=").append(tag);
            }
            if (methodParameterTypes != null && !methodParameterTypes.isEmpty()) {
                sb.append(", methodParameterTypes=").append(methodParameterTypes);
            }
            if (typeVariables != null && !typeVariables.isEmpty()) {
                sb.append(", typeVariables=").append(typeVariables);
            }
            if (isInterceptor) {
                sb.append(", isInterceptor=").append(isInterceptor);
            }
            sb.append(']');
            return sb.toString();
        }
    }

    record AnnotatedComponent(TypeMirror type, TypeElement typeElement, @Nullable String tag, ExecutableElement constructor, List<TypeMirror> methodParameterTypes,
                              List<TypeMirror> typeVariables, boolean isInterceptor) implements ComponentDeclaration {
        @Override
        public Element source() {
            return this.constructor;
        }

        @Override
        public String declarationString() {
            return typeElement.getQualifiedName().toString();
        }

        @Override
        public String toString() {
            final StringBuilder sb = new StringBuilder("AnnotatedComponent[");
            sb.append("type=").append(type);
            sb.append(", typeElement=").append(typeElement);
            sb.append(", constructor=").append(constructor);
            if (tag != null) {
                sb.append(", tag=").append(tag);
            }
            if (methodParameterTypes != null && !methodParameterTypes.isEmpty()) {
                sb.append(", methodParameterTypes=").append(methodParameterTypes);
            }
            if (typeVariables != null && !typeVariables.isEmpty()) {
                sb.append(", typeVariables=").append(typeVariables);
            }
            if (isInterceptor) {
                sb.append(", isInterceptor=").append(isInterceptor);
            }
            sb.append(']');
            return sb.toString();
        }
    }

    record DiscoveredAsDependencyComponent(DeclaredType type, TypeElement typeElement, ExecutableElement constructor, @Nullable String tag) implements ComponentDeclaration {

        @Override
        public Element source() {
            return this.constructor;
        }

        @Override
        public boolean isTemplate() {
            return false;
        }

        @Override
        public boolean isInterceptor() {
            return false;
        }

        @Override
        public String declarationString() {
            return typeElement.getQualifiedName().toString();
        }

        @Override
        public String toString() {
            final StringBuilder sb = new StringBuilder("DiscoveredAsDependencyComponent[");
            sb.append("type=").append(type);
            sb.append(", typeElement=").append(typeElement);
            sb.append(", constructor=").append(constructor);
            if (tag != null) {
                sb.append(", tag=").append(tag);
            }
            sb.append(']');
            return sb.toString();
        }
    }

    record FromExtensionComponent(
        TypeMirror type,
        Element source,
        List<TypeMirror> dependencyTypes,
        List<String> dependencyTags,
        @Nullable String tag,
        Function<CodeBlock, CodeBlock> generator
    ) implements ComponentDeclaration {
        @Override
        public boolean isInterceptor() {
            return false;
        }

        @Override
        public String declarationString() {
            return source.getEnclosingElement().toString() + "." + source.getSimpleName();
        }

        @Override
        public String toString() {
            final StringBuilder sb = new StringBuilder("FromExtensionComponent[");
            sb.append("type=").append(type);
            sb.append(", source=").append(source);
            if (tag != null) {
                sb.append(", tag=").append(tag);
            }
            if (dependencyTypes != null && !dependencyTypes.isEmpty()) {
                sb.append(", dependencyTypes=").append(dependencyTypes);
            }
            if (dependencyTags != null && dependencyTags.stream().anyMatch(Objects::nonNull)) {
                sb.append(", dependencyTags=").append(dependencyTags);
            }
            sb.append(']');
            return sb.toString();
        }
    }

    record PromisedProxyComponent(TypeElement typeElement, TypeMirror type, com.palantir.javapoet.ClassName className) implements ComponentDeclaration {
        public PromisedProxyComponent(TypeElement typeElement, com.palantir.javapoet.ClassName className) {
            this(typeElement, typeElement.asType(), className);
        }

        public PromisedProxyComponent withType(TypeMirror type) {
            return new PromisedProxyComponent(typeElement, type, className);
        }


        @Override
        public Element source() {
            return this.typeElement;
        }

        @Override
        public String tag() {
            return CommonClassNames.promisedProxy.canonicalName();
        }

        @Override
        public boolean isInterceptor() {
            return false;
        }

        @Override
        public String declarationString() {
            return "<Proxy>";
        }
    }

    record OptionalComponent(TypeMirror type, @Nullable String tag) implements ComponentDeclaration {
        @Override
        public Element source() {
            return null;
        }

        @Override
        public boolean isInterceptor() {
            return false;
        }

        @Override
        public String declarationString() {
            return "<EmptyOptional>";
        }

        @Override
        public String toString() {
            final StringBuilder sb = new StringBuilder("OptionalComponent[");
            sb.append("type=").append(type);
            if (tag != null) {
                sb.append(", tag=").append(tag);
            }
            sb.append(']');
            return sb.toString();
        }
    }

    static ComponentDeclaration fromModule(ProcessingContext ctx, ModuleDeclaration module, ExecutableElement method) {
        var type = method.getReturnType();
        if (TypeParameterUtils.hasRawTypes(type)) {
            throw new ProcessingErrorException("Components with raw types can break dependency resolution in unpredictable way so they are forbidden", method);
        }
        var tags = TagUtils.parseTagValue(method);
        var parameterTypes = method.getParameters().stream().map(VariableElement::asType).toList();
        var typeParameters = method.getTypeParameters().stream().map(TypeParameterElement::asType).toList();
        var isInterceptor = ctx.serviceTypeHelper.isInterceptor(type);
        return new FromModuleComponent(type, module, tags, method, parameterTypes, typeParameters, isInterceptor);
    }

    static ComponentDeclaration fromAnnotated(ProcessingContext ctx, TypeElement typeElement) {
        var constructors = CommonUtils.findConstructors(typeElement, m -> m.contains(Modifier.PUBLIC));
        if (constructors.size() != 1) {
            throw new ProcessingErrorException("Components with raw types can break dependency resolution in unpredictable way so they are forbidden", typeElement);
        }
        var constructor = constructors.get(0);
        var type = typeElement.asType();
        if (TypeParameterUtils.hasRawTypes(type)) {
            ctx.messager.printMessage(Diagnostic.Kind.WARNING, "Components with raw types can break dependency resolution in unpredictable way", typeElement);
        }
        var tags = TagUtils.parseTagValue(typeElement);
        var parameterTypes = constructor.getParameters().stream().map(VariableElement::asType).toList();
        var typeParameters = typeElement.getTypeParameters().stream().map(TypeParameterElement::asType).toList();
        var isInterceptor = ctx.serviceTypeHelper.isInterceptor(type);
        return new AnnotatedComponent(type, typeElement, tags, constructor, parameterTypes, typeParameters, isInterceptor);
    }

    static ComponentDeclaration fromDependency(ProcessingContext ctx, TypeElement typeElement, DeclaredType declaredType) {
        var constructors = CommonUtils.findConstructors(typeElement, m -> m.contains(Modifier.PUBLIC));
        if (constructors.size() != 1) {
            throw new ProcessingErrorException("Can't create component from discovered as dependency class: class should have exactly one public constructor", typeElement);
        }
        var constructor = constructors.get(0);
        if (TypeParameterUtils.hasRawTypes(declaredType)) {
            throw new ProcessingErrorException("Components with raw types can break dependency resolution in unpredictable way so they are forbidden", typeElement);
        }
        var tags = TagUtils.parseTagValue(typeElement);
        return new DiscoveredAsDependencyComponent(declaredType, typeElement, constructor, tags);
    }

    static ComponentDeclaration fromExtension(ProcessingContext ctx, ExtensionResult.GeneratedResult generatedResult) {
        var sourceMethod = generatedResult.sourceElement();
        if (sourceMethod.getKind() == ElementKind.CONSTRUCTOR) {
            var parameterTypes = sourceMethod.getParameters().stream().map(VariableElement::asType).toList();
            var parameterTags = sourceMethod.getParameters().stream().map(TagUtils::parseTagValue).toList();
            var typeElement = (TypeElement) sourceMethod.getEnclosingElement();
            var tag = TagUtils.parseTagValue(sourceMethod);
            if (tag == null) {
                tag = TagUtils.parseTagValue(typeElement);
            }
            var type = typeElement.asType();
            if (TypeParameterUtils.hasRawTypes(type)) {
                throw new ProcessingErrorException("Components with raw types can break dependency resolution in unpredictable way so they are forbidden", sourceMethod);
            }
            var className = ClassName.get(typeElement);

            return new FromExtensionComponent(type, sourceMethod, parameterTypes, parameterTags, tag, dependencies -> typeElement.getTypeParameters().isEmpty()
                ? CodeBlock.of("new $T($L)", className, dependencies)
                : CodeBlock.of("new $T<>($L)", className, dependencies));
        } else {
            var type = generatedResult.targetType().getReturnType();
            var parameterTypes = generatedResult.targetType().getParameterTypes();
            if (TypeParameterUtils.hasRawTypes(type)) {
                throw new ProcessingErrorException("Components with raw types can break dependency resolution in unpredictable way so they are forbidden", sourceMethod);
            }
            var parameterTags = sourceMethod.getParameters().stream().map(TagUtils::parseTagValue).toList();
            var tag = TagUtils.parseTagValue(sourceMethod);
            var typeElement = (TypeElement) sourceMethod.getEnclosingElement();
            var className = ClassName.get(typeElement);
            return new FromExtensionComponent(type, sourceMethod, new ArrayList<>(parameterTypes), parameterTags, tag, dependencies -> CodeBlock.of("$T.$N($L)", className, sourceMethod.getSimpleName(), dependencies));
        }
    }

    static ComponentDeclaration fromExtension(ExtensionResult.CodeBlockResult generatedResult) {
        return new FromExtensionComponent(
            generatedResult.componentType(),
            generatedResult.source(),
            generatedResult.dependencyTypes(),
            generatedResult.dependencyTags(),
            generatedResult.componentTag(),
            generatedResult.codeBlock()
        );
    }
}

File: kora-app-symbol-processor/src/main/kotlin/ru/tinkoff/kora/kora/app/ksp/declaration/ComponentDeclaration.kt
package ru.tinkoff.kora.kora.app.ksp.declaration

import com.google.devtools.ksp.closestClassDeclaration
import com.google.devtools.ksp.isConstructor
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSDeclaration
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.google.devtools.ksp.symbol.KSTypeArgument
import com.google.devtools.ksp.symbol.KSTypeParameter
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.MemberName
import com.squareup.kotlinpoet.TypeName
import com.squareup.kotlinpoet.ksp.toClassName
import ru.tinkoff.kora.kora.app.ksp.ProcessingContext
import ru.tinkoff.kora.kora.app.ksp.extension.ExtensionResult
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findAnnotation
import ru.tinkoff.kora.ksp.common.CommonClassNames
import ru.tinkoff.kora.ksp.common.KspCommonUtils.fixPlatformType
import ru.tinkoff.kora.ksp.common.TagUtils
import ru.tinkoff.kora.ksp.common.TagUtils.parseTags
import ru.tinkoff.kora.ksp.common.exception.ProcessingErrorException

sealed interface ComponentDeclaration {
    val type: KSType
    val source: KSDeclaration
    val tags: Set<String>

    fun declarationString(): String

    fun isTemplate(): Boolean {
        for (argument in type.arguments) {
            if (argument.hasGenericVariable()) {
                return true
            }
        }
        return false
    }

    fun isDefault(): Boolean {
        return false
    }

    data class FromModuleComponent(
        override val type: KSType,
        val module: ModuleDeclaration,
        override val tags: Set<String>,
        val method: KSFunctionDeclaration,
        val methodParameterTypes: List<KSType>,
        val typeVariables: List<KSTypeArgument>
    ) : ComponentDeclaration {
        override val source get() = this.method
        override fun declarationString() = module.element.qualifiedName?.asString() + "." + method.simpleName.asString()

        override fun isDefault(): Boolean {
            return method.findAnnotation(CommonClassNames.defaultComponent) != null
        }
    }

    data class AnnotatedComponent(
        override val type: KSType,
        val classDeclaration: KSClassDeclaration,
        override val tags: Set<String>,
        val constructor: KSFunctionDeclaration,
        val methodParameterTypes: List<KSType>,
        val typeVariables: List<KSTypeArgument>
    ) : ComponentDeclaration {
        override val source get() = this.constructor
        override fun declarationString() = classDeclaration.qualifiedName?.asString().toString()
    }

    data class DiscoveredAsDependencyComponent(
        override val type: KSType,
        val classDeclaration: KSClassDeclaration,
        val constructor: KSFunctionDeclaration,
        override val tags: Set<String>
    ) : ComponentDeclaration {
        override val source get() = this.constructor
        override fun declarationString() = classDeclaration.qualifiedName?.asString().toString()
    }

    data class FromExtensionComponent(
        override val type: KSType,
        override val source: KSDeclaration,
        val methodParameterTypes: List<KSType>,
        val methodParameterTags: List<Set<String>>,
        override val tags: Set<String>,
        val generator: (CodeBlock) -> CodeBlock
    ) : ComponentDeclaration {
        override fun declarationString(): String {
            return source.parentDeclaration?.qualifiedName?.asString().toString() + source.simpleName.asString()
        }

    }

    data class PromisedProxyComponent(
        override val type: KSType,
        val classDeclaration: KSClassDeclaration,
        val className: TypeName

    ) : ComponentDeclaration {
        override val source get() = this.classDeclaration
        override val tags get() = setOf(CommonClassNames.promisedProxy.canonicalName)
        override fun declarationString() = "<Proxy>"
    }


    data class OptionalComponent(
        override val type: KSType,
        override val tags: Set<String>
    ) : ComponentDeclaration {
        override val source get() = type.declaration
        override fun declarationString() = "Optional.empty"
    }


    companion object {
        fun fromModule(ctx: ProcessingContext, module: ModuleDeclaration, method: KSFunctionDeclaration): FromModuleComponent {
            // modules can be written in java so we better fix platform nullability
            val type = method.returnType!!.resolve().fixPlatformType(ctx.resolver)
            if (type.isError) {
                throw ProcessingErrorException("Component type is not resolvable in the current round of processing", method)
            }
            val tags = TagUtils.parseTagValue(method)
            val parameterTypes = method.parameters.map { it.type.resolve().fixPlatformType(ctx.resolver) }
            val typeParameters = method.typeParameters.map {
                val t = it.bounds.firstOrNull()?.resolve()?.fixPlatformType(ctx.resolver) ?: ctx.resolver.builtIns.anyType

                ctx.resolver.getTypeArgument(
                    ctx.resolver.createKSTypeReferenceFromKSType(t),
                    it.variance
                )
            }
            return FromModuleComponent(type, module, tags, method, parameterTypes, typeParameters)
        }

        fun fromAnnotated(ctx: ProcessingContext, classDeclaration: KSClassDeclaration): AnnotatedComponent {
            val constructor = classDeclaration.primaryConstructor
            if (constructor == null) {
                throw ProcessingErrorException("@Component annotated class should have primary constructor", classDeclaration)
            }
            val typeParameters = classDeclaration.typeParameters.map {
                val t = it.bounds.firstOrNull()?.resolve() ?: ctx.resolver.builtIns.anyType

                ctx.resolver.getTypeArgument(
                    ctx.resolver.createKSTypeReferenceFromKSType(t),
                    it.variance
                )
            }
            val type = classDeclaration.asType(listOf())
            val tags = TagUtils.parseTagValue(classDeclaration)
            val parameterTypes = constructor.parameters.map { it.type.resolve() }

            return AnnotatedComponent(type, classDeclaration, tags, constructor, parameterTypes, typeParameters)
        }

        fun fromDependency(@Suppress("UNUSED_PARAMETER") ctx: ProcessingContext, classDeclaration: KSClassDeclaration, type: KSType): DiscoveredAsDependencyComponent {
            val constructor = classDeclaration.primaryConstructor
            if (constructor == null) {
                throw ProcessingErrorException("No primary constructor to parse component for: $classDeclaration", classDeclaration)
            }
            if (type.isError) {
                throw ProcessingErrorException("Component type is not resolvable in the current round of processing", classDeclaration)
            }
            val tags = TagUtils.parseTagValue(classDeclaration)

            return DiscoveredAsDependencyComponent(type, classDeclaration, constructor, tags)
        }

        fun fromExtension(ctx: ProcessingContext, extensionResult: ExtensionResult.GeneratedResult): FromExtensionComponent {
            val sourceMethod = extensionResult.constructor
            val sourceType = extensionResult.type
            val parameterTypes = sourceType.parameterTypes.map { it!!.fixPlatformType(ctx.resolver) }
            val parameterTags = sourceMethod.parameters.map { it.parseTags() }
            val type = sourceType.returnType!!
            if (type.isError) {
                throw ProcessingErrorException("Component type is not resolvable in the current round of processing", sourceMethod)
            }
            val tag = if (sourceMethod.isConstructor()) {
                sourceMethod.closestClassDeclaration()!!.parseTags()
            } else {
                sourceMethod.parseTags()
            }

            return FromExtensionComponent(type, sourceMethod, parameterTypes, parameterTags, tag) {
                if (sourceMethod.isConstructor()) {
                    val clazz = sourceMethod.closestClassDeclaration()!!
                    CodeBlock.of("%T(%L)", clazz.toClassName(), it)
                } else {
                    val parent = sourceMethod.parentDeclaration
                    if (parent is KSClassDeclaration) {
                        CodeBlock.of("%M(%L)", MemberName(parent.toClassName(), sourceMethod.simpleName.asString()), it)
                    } else {
                        CodeBlock.of("%M(%L)", MemberName(sourceMethod.packageName.asString(), sourceMethod.simpleName.asString()), it)
                    }
                }
            }
        }

        fun fromExtension(extensionResult: ExtensionResult.CodeBlockResult): FromExtensionComponent {
            return FromExtensionComponent(
                extensionResult.componentType,
                extensionResult.source,
                extensionResult.dependencyTypes,
                extensionResult.dependencyTags,
                extensionResult.componentTag,
                extensionResult.codeBlock
            )
        }
    }
}

private fun KSTypeArgument.hasGenericVariable(): Boolean {
    val type = this.type
    if (type == null) {
        return true
    }
    val resolvedType = type.resolve()
    if (resolvedType.declaration is KSTypeParameter) {
        return true
    }

    for (param in resolvedType.arguments) {
        if (param.hasGenericVariable()) {
            return true
        }
    }
    return false
}
